### PR TITLE
[AK/Bitmap, Heap/kmalloc] optimisations for x86 cpus

### PR DESCRIPTION
**Pull request by nimelehin**
Kmalloc is more efficient now, because of this improvement the load
of CPU is less in all scenarios

Makefile is updated with new GCC parameter -mlzcnt

### CODE REVIEW

**Comment by awesomekling** [comment]
How much faster is this? What is the benchmark?

**Comment by nimelehin** [comment]
Hello. I just put kmalloc in syscall_handler and was waiting when memory was out. And when memory is out I compared CPU's ticks (in both cases) and performance boost is about 25%.

**Comment by awesomekling** [comment]
That sounds like an extremely specific microbenchmark. How about a macrobenchmark like compiling a "Hello world" program with gcc? Also 25% faster than what? What are the cycle counts  before and after?

I'm asking because you're adding complexity without proportional explanation or motivation. :)

Also please observe the contribution guidelines: https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

Since this is a C++ project and not a C project, can we do this in a more C++ way? Can we instantiate an `AK::Bitmap` wrapper around the `alloc_map` and create helper functions in `Bitmap` that provide this functionality so we can write the `kmalloc` logic in a more expressive way?

It would be nice if the logic looked more like this (example below), and we could put the cleverness in the `Bitmap` class instead:

```cpp
Bitmap bitmap = Bitmap::wrap(alloc_map, POOL_SIZE / CHUNK_SIZE);
Optional<size_t> first_fit = bitmap.find_first_fit(chunks_needed);
if (!first_fit.has_value())
    panic("kmalloc heap is full :(");
bitmap.set_range(first_fit, chunks_needed, true);
...
```

**Comment by nimelehin** [comment]
In tests 3 program took part:
1. Current kmalloc implementation (master)
2. Kmalloc implementation with increased bucket size to 32 bits (presented in my commit as implementation for non-x86 cpus)
3. Kmalloc implementation with lzcnt (presented in my commit as implementation for x86 cpus)

To calculate the performance boost the several tests were run:

### Test 1: Running GCC

There were 2 different configurations: the first with CHUNK_SIZE=32 and the second with CHUNK_SIZE=4 (to check the performance when there are a lot of taken chunks)

**Result 1.1: CHUNK_SIZE=32**

	Time of compilation (for each implementations) (ms)

	Implementation 1	
	[3304, 3301, 3218, 3238, 3273, 3277, 3256, 3236, 3189, 3236, 3355]
	Avg: 3262

	Implementation 2
	[3323, 3177, 3243, 3130, 3150, 3122, 3139, 3188, 3150, 3141, 3186]
	Avg: 3177

	Implementation 3
	[3309, 3184, 3253, 3320, 3257, 3162, 3133, 3137, 3195, 3144, 3233]
	Avg: 3211

**Conclusion:**
* The performance is about the same in all 3 cases. So when the alloc_map is quite empty (there a lot place to allocate) there is no leader.



**Result 1.2: CHUNK_SIZE=8**
	Time of compilation (for each implementations) (ms)

	Implementation 1	
	[4491, 4283, 4342, 4483, 4471, 4449, 4487, 4511, 4453, 4573, 4527, 4583]
	Avg: 4471

	Implementation 2
	[4236, 4126, 4225, 4226, 4221, 4307, 4252, 4264, 4216, 4324, 4386, 4361]
	Avg: 4262

	Implementation 3
	[3854, 3686, 3747, 3746, 3728, 3829, 3734, 3708, 3831, 3759, 3738, 3755]
	Avg: 3759
	
	The 3rd implementation (with lzcnt) is 19% more efficient than the 1st (master). 
	The 3rd implementation (with lzcnt) is 13% more efficient than the 2nd (32bit non-x86). 
	The 2nd implementation (32bit non-x86) is 5% more efficient than the 1st (master). 
	
**Conclusion:**
* When there are a lot of taken chunks in the alloc_map implementations 2 and 3 are more efficient than the current. The most effective implementation is 3rd.



### Test 2: Calling kmalloc while there is some space

We are going to call kmalloc while there is space to allocate. The sooner all space is taken the faster kmalloc is.

**Result 2:** 
	Ticks of cpu until there is no memory (since the cpu has started)
	
	Implementation 1
	[41774019579,
	42031540792,
	41439121749,
	42059286884,
	41814278016]
	Avg: 41823649404

	Implementation 2
	[33787471086,
	33300001473,
	31884552015,
	33669089462,
	31673878635]
	Avg: 32862998534

	Implementation 3
	[31894981346,
	33544974641,
	31653914648,
	32286556015,
	32288015759]
	Avg: 32333688481

	The 3rd implementation (with lzcnt) is 30% more efficient than the 1st (master). 
	The 3rd implementation (with lzcnt) is 1.5% more efficient than the 2nd (32bit non-x86). 
	The 2nd implementation (32bit non-x86) is 27% more efficient than the 1st (master). 
	
**Conclusion:**
* Implementations 2 and 3 are more efficient than the current. Based on Test 1 I think the performance become so big when the alloc_map is full (at the end of the test)
* The difference between 3rd and 2nd implementation isn’t as big as it’s in Test 1.2 because almost all buckets are skipped (because they are fully taken), so there are not many times when lzcnt are called to show the boost which it can give.


### Report Conclusion:
* The 32 bit implementations (2 and 3) are more efficient when the alloc_map is full and about the same performance when the alloc_map is not full. 
* The 3rd implementation is the most efficient in Test 2 and Test 1.2 and has the same performance in Test 1.1.

**Comment by awesomekling** [comment]
Okay thank you for testing. Numbers look good :)

Two issues:

* Let's be C++ programmers and make this a reusable feature of the `Bitmap` class instead of doing everything in kmalloc.
* `LZCNT` is a relatively recent instruction, only available on CPU's since 2011. Do we need to do some `CPUID` feature detection and enable/disable this code path at runtime?

**Comment by nimelehin** [comment]
I got, I will make it more C++)

If CPU doesn’t support lzcnt, it will be executed as BSR instruction. So there is the only case to check: when the operand is zero (bsr and lzcnt produce different outputs)

File: AK/Bitmap.h [hunk 1]
```diff
@@ -171,14 +171,14 @@ class Bitmap {
         if (!first_index.has_value())
             return {};
 
-        size_t free_region_start = first_index;
-        size_t free_region_size = 1;
+        size_t free_region_start = 0;
+        size_t free_region_size = 0;
```
Comment (awesomekling) [inline AK/Bitmap.h]: Hmm, why is this logic being changed? Should it not be like this:
```suggestion
        size_t free_region_start = first_index.value();
        size_t free_region_size = 1;
```

Comment (nimelehin) [reply AK/Bitmap.h]: I did change the logic. I thought it would be a little bit cleaner, because the loop will set the right values in the first if.

If it’s better to return everything on its place, I will)

Comment (awesomekling) [reply AK/Bitmap.h]: But the loop doesn't start at `0`, it starts at `first_index + 1`, so it may never find the same values.

Also you are changing the loop start from `first_index + 1` to `first_index`. The point of this code is that we've already found the first index in the call to `find_first_unset()` above and we want to continue accumulating *after* the first unset bit.

**Comment by awesomekling** [comment]
Please squash fixups into the previous commit instead of adding changes on top of changes :)

**Comment by nimelehin** [comment]
The work isn’t finished. I am working on it)

**Comment by nimelehin** [comment]
The new [set of tests](https://docs.google.com/spreadsheets/d/1lHWXjPVXLbQ0Nkvd7XZCMFBcKm879ObXXdymP6OODbM/edit?usp=sharing))
Actually we lose some performance with `CPUID` but not too much and almost in all tests solutions with `lzcnt` are better. But there is a test 5 and it's designed to be the worst for lzcnt solutions. Need to say that sometimes (only in test 5) the current master solution is even better. Actually the test 5 is when we have a bitmap with a pattern like (01010101...). I think it's possible to fix it adding `SlabAllocator` features to `kmalloc`: so when there is a small range to allocate, it's better to use `SlabAllocator`, since it's faster and can resolve the situation above. Actually maybe it's a good idea to use `best fit` instead of `first fit` for big allocations

**Comment by awesomekling** [comment]
Are you doing `cpuid` every time? We should only do it once on startup and then remember the value :)

**Comment by nimelehin** [comment]
No, not every time.

Now I use this, and when in a constructor of class `__builtin_cpu_init()`
`#define leading_zeroes_cnt(x) (__builtin_cpu_supports("popcnt") ? __builtin_ia32_lzcnt_u32(x) : 31 -  __builtin_ia32_lzcnt_u32(x));`

Before that I used `cpuid` and put result in a variable. Actually performance was about the same, but I think with __builins it's more readable)

**Comment by nimelehin** [comment]
It seems like it's ready. I moved from `lzcnt` to `bsf`, so now there is no `cpuid` and speeds are as they have been in the first test.

I did some rebase&squash, say if it's okay)

**Comment by awesomekling** [comment]
Please rebase on top of master to get rid of all merge commits.
Also make sure to squash commits that iterate on the same code.

**Comment by nimelehin** [comment]
Now commits look nice

File: AK/Bitmap.h [hunk 2]
```diff
@@ -203,6 +203,32 @@ class Bitmap {
         return first_index;
     }
 
+    Optional<size_t> find_first_fit(size_t req_length) const
```
Comment (awesomekling) [inline AK/Bitmap.h]: Please use full words in parameter names. How about `minimum_length`?

File: AK/Bitmap.h [hunk 3]
```diff
@@ -33,6 +33,8 @@
 #include <AK/Types.h>
 #include <AK/kmalloc.h>
 
+#define BITMAP_SIZE_MAX 0xffffffff
```
Comment (awesomekling) [inline AK/Bitmap.h]: Please use `constexpr` variables instead of macros.

File: AK/Bitmap.h [hunk 4]
```diff
@@ -163,69 +157,151 @@ class Bitmap {
...
+#else
+        auto trailing_zeroes_cnt = [](u32 val) -> u32{
+            for (u8 i = 0; i < 32; ++i) {
+                if ((val >> i) & 1) {
+                    return i;
+                }
+            }
+            return 0;
+        };
+#endif /* AK_ARCH_I386 */
```
Comment (awesomekling) [inline AK/Bitmap.h]: Please put this in `AK/Platform.h`.
Also, please don't abbreviate random words. `trailing_zeroes_count` or `count_trailing_zeroes`.

File: AK/Bitmap.h [hunk 5]
```diff
@@ -163,69 +157,151 @@ class Bitmap {
...
+     *        will be changed and new value is the offset of the
+     *        found block
+     * @min_length: minimum size of the range which will be returned
+     * @max_length: shows when to stop adding unset bits to the range.
+     *            This is used for optimisation, since the range of
+     *            unset bits can be long, and we don't need the whole
+     *            size, so we can stop when we've reached @max_length.
+     * It's recomended to use @min_length and @max_length, since they
+     * give a boost.
+     */
```
Comment (awesomekling) [inline AK/Bitmap.h]: Please use C++ `//` comments.
The function is not *supposed* to find the next range, it *will* find it :)

File: AK/Bitmap.h [hunk 6]
```diff
@@ -163,69 +157,151 @@ class Bitmap {
...
-                if (free_region_size == 0)
-                    free_region_start = j;
-                free_region_size++;
-            } else {
-                if (max_region_size < free_region_size) {
-                    max_region_size = free_region_size;
-                    max_region_start = free_region_start;
+        u32 *bitmap32 = (u32*)m_data;
+        
+        // calculating the start offset
```
Comment (awesomekling) [inline AK/Bitmap.h]: [Please write comments in proper English.](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments)

File: AK/Bitmap.h [hunk 7]
```diff
@@ -163,69 +157,151 @@ class Bitmap {
...
+        size_t start_bucket_id = from / 32;
+        size_t init_offset_in_bucket = from % 32;
+        
+        size_t *start_of_free_chunks = &from; // this value points to @from, since @from will be changed
+        size_t free_chunks = 0;
+        
+        for (size_t buc_id = start_bucket_id; buc_id < m_size / 32; ++buc_id) {
```
Comment (awesomekling) [inline AK/Bitmap.h]: Please use full words for variable names. Also, these are not ID's but indexes.

File: AK/Bitmap.h [hunk 8]
```diff
@@ -163,69 +157,151 @@ class Bitmap {
...
+            auto length_of_found_range = find_next_range_of_unset_bits(start, max_region_size + 1, max_length);
+            if (length_of_found_range.has_value()) {
+                // if we have value, it's definitely begger than current
+                max_region_start = start;
+                max_region_size = length_of_found_range.value();
+                start += max_region_size;
             } else {
-                free_region_start = 0;
-                free_region_size = 0;
+                // there are no ranges whach are bigger than current, so let's finish
```
Comment (awesomekling) [inline AK/Bitmap.h]: s/whach/which/

File: Kernel/Heap/kmalloc.cpp [hunk 9]
```diff
@@ -109,6 +111,22 @@ void* kmalloc_page_aligned(size_t size)
     return ptr;
 }
 
+inline void* kmalloc_allocate(size_t first_chunk, size_t chunks_needed) {
```
Comment (awesomekling) [inline Kernel/Heap/kmalloc.cpp]: Please use `clang-format` to ensure that your code is properly formatted.

File: AK/Bitmap.h [hunk 10]
```diff
@@ -275,24 +277,25 @@ class Bitmap {
...
         }
 
         found_range_size = max_region_size;
-        if (!max_region_size) {
-            return {};
+        if (max_region_size) {
+            return max_region_start;
         }
-        return max_region_start;
+        return {};
```
Comment (awesomekling) [inline AK/Bitmap.h]: There are all kinds of unrelated changes in this commit. (Unrelated to the commit message.)
Please squash them into the relevant previous commit instead so that commits are small and precise.

File: Kernel/Heap/kmalloc.cpp [hunk 11]
```diff
@@ -41,6 +41,7 @@
 #include <LibBareMetal/StdLib.h>
 
 #define SANITIZE_KMALLOC
+#define KMALLOC_BEST_FIT_BOUNDARY 128
```
Comment (awesomekling) [inline Kernel/Heap/kmalloc.cpp]: The commit message does not mention anything about this?

File: AK/Bitmap.h [hunk 12]
```diff
@@ -163,68 +158,141 @@ class Bitmap {
...
+#ifndef count_trailing_zeroes
+        auto count_trailing_zeroes = [](u32 val) -> u32{
+            for (u8 i = 0; i < 32; ++i) {
+                if ((val >> i) & 1) {
+                    return i;
+                }
+            }
+            return 0;
+        };
+#endif // count_trailing_zeroes
```
Comment (awesomekling) [inline AK/Bitmap.h]: The point of moving this to Platform.h was to get rid of the inline definition here :)

**Comment by nimelehin** [comment]
Hello. I'm sorry for being away for long) I made some changes. Thank you for your help and suggestions)

**Comment by awesomekling** [comment]
Merged manually in c0a4cf5e8d78f8a24551c2e98219d0c6eff4aefb :)
